### PR TITLE
Keysig fixes

### PR DESCRIPTION
**Pull request by sammik**
Resolves: #14274
Resolves: #17547
Resolves: #17508
Resolves: #17506
Resolves: #17498
Resolves: #17493
Resolves: #17407
Resolves: #16983
Resolves: #17391
Resolves: #17375
Resolves: #17293
Resolves: #17613
Resolves: #17642
Resolves: #17626
Resolves: #17661
Resolves: #17665 
Resolves: #17732 

This PR adds "concertKey" to key definition. It is relevant for transposing instruments, to concert key be unchanged, when switch between transposing and concert pitch.
Now we have "key", which is actual key, shown in score (transposing or concert), and "concert key", which holds concert key information.

Keysig is nowstored as
```
<KeySig>
    <c_accidental>0</c_accidental>
</KeySig>
```
with optional `<t_accidental>2</t_accidental>` if needed

PR also corrects note spellings in transposing pitch, to reflect key signature. And also Harmony. 

PR adds default option for prefer sharps or flats, to be "less", which means, program prefer 5 sharps ower 7 flats, or 5 flats ower 7 sharps.

Most of the issues are addresed and fixed by separate commits (but not all).

https://github.com/musescore/MuseScore/assets/1646034/f926eb2a-a3ea-4127-ba67-6be846c63cb4

- [x] I signed the [CLA](https://musescore.org/en/cla)
- [x] The title of the PR describes the problem it addresses
- [x] Each commit's message describes its purpose and effects, and references the issue it resolves
- [x] If changes are extensive, there is a sequence of easily reviewable commits
- [x] The code in the PR follows [the coding rules](https://github.com/musescore/MuseScore/wiki/CodeGuidelines)
- [x] There are no unnecessary changes
- [x] The code compiles and runs on my machine, preferably after each commit individually
- [ ] I created a unit test or vtest to verify the changes I made (if applicable)

TODO 
~- add warning - see #17626~ possible conflict with #17551 
- needs translation? (preferSharpFlat)

### CODE REVIEW

**Comment by oktophonie** [comment]
Amazing!
I can confirm that all the listed issues seem to be fixed here (I won't pretend to have tested each one thoroughly, but in the example cases and other similar ones, all seems fine).

There are still a couple of problems, which I uncovered while testing the fix for #17391. To  summarise the situation as I understand it:

If we select a bar where there is a 'normal' key signature change  - that is, not on that arises because of an instrument change - and insert a bar before it (Add > Bars > Insert before selection, or Ctrl+Ins), the key signature change stays where it is, i.e. the new bar is inserted 'between' the key signature and the selected bar. In other words, this:
image
becomes this:
image
This is the case before and after this PR and is presumably by design. (I can see an argument for doing it the other way, and having the inserted bar 'before' the key signature - but we have no way of knowing which the user would prefer in any given case.)

In the case of an instrument change, before this PR, adding a bar before this one:
image
would result in this:
image
which seems "consistent", but is incorrect, because the key signature change here only arises because of the instrument change, so they should happen at the same point.

With this PR the result is this:
image
which is correct.

There are still some problems, however.
Let's say I _did_ want the instrument change, and therefore also the key signature change, to be a bar earlier. No problem, I can cut and paste it to that bar. Unfortunately, if I do so, the key signature does not rematerialise:
image
To be fair, that is a pre-existing problem, not caused by this PR, but is perhaps in scope to be fixed here?

More problematic: if there's a simultaneous key change _and_ instrument change:
image
if we insert a bar before that point we get this:
image
where the upper stave should get the key signature change in bar 2, but it goes missing.

This can be reproduced more directly by just deleting the instrument change. This:
image
becomes this:
image
(the key signature changes in these examples are definitely global, not local).

**Comment by sammik** [comment]
To be clear, correct behaviour would be following?

behaviour

**Comment by Jojo-Schmitz** [comment]
Isn't this PR changing the score file format in a way that would require an update to file format version number?
What happpens with such scores when opened with 4.0.2?

**Comment by cbjeukendrup** [comment]
The version number change is going to be there anyway (for sure), so we don't need to worry. 

(We only need to worry about scores that are created in the current master builds, because those don't have the new number, but do have the new format. But that's not this PR's problem, that's a general issue.)

**Comment by sammik** [comment]
I investigate a bit.

This problem with delete instrument change is not caused by this PR, and is present in 3.6.2 too.
It depends, if key signature was there befere user adds instrument change, or was added later.

I see two general problems / questions
1. we dont know, if key signature is local, or global, there is no property "local" (I think, it could be relevant information and useful property to add)
2. what is property "forInstrumentChange" useful for? Does we even need it?

**Comment by sammik** [comment]
OK, I thought about it. I will try to do it so:

If only reason, why key sig is there is instrument change (no global, or local key sig present at that place), keysig would be marked "forInstrumentChange". Otherwise not.

And if keysig is marked for instrument change, it wouldnt move if user insert measure, and will be removed, when user removes instrument change.
Otherwise, it will be translated (and moved if inserting measure).

**Comment by oktophonie** [comment]
@sammik if you're going to address the issues in my above comment separately (https://github.com/musescore/MuseScore/issues/17613) could you rebase this one and attend to the utests so we can look to merge it?

**Comment by sammik** [comment]
@oktophonie I think, better to include it here. 
There are still some issues with instrumentChange, I like to solve all in one. Working on it right now.

**Comment by sammik** [comment]
I corrected something, see video.

About instrument change key signatures.

If normal key signature already is on place of instrument change, or is added to place of instrument change, it should behave as ordinary key signature (if precedent ks is changed, ic ks remain unchanged, ...)

If instrument change is removed, ic ks is removed too. If there were normal ks, it remains and is transposed.

If user insert measure, icks is not moved. If normal ks was there, it is moved and in place of instrument change, ic ks is generated.


https://github.com/musescore/MuseScore/assets/1646034/00d0c6ed-6601-4aa1-9ccf-3ff06f199104

I am not sure about code application. I choosed to use tag for ks "ForInstrumentalChange::NONE", if there is no instrumental change, "ForInstrumentalChange::PASIVE" if ic is there, abut also ks, and "ForInstrumentalChange::ACTIVE" if ic there and no additional ks there (ekvivalent of previous "forInstrumenChange true"
I thought, it is less difficualt, to have this info (instrument change) inside ks, so we doesnt need to check, if instrument change is there or not when deleting ks, or inserting measure, ...

But still one big proble remains.
If opening older files, and instrument change is there, we need to set correct concertKey. But we dont know it, because key signatures are readed before instrument change text, so transposition information is missing.

I tried something, but unsuccesfull.

I think, logich should be: if read older files "accidental" tag - tread.cpp line 1238, we need to check, if instrument change is not on that place, and if it is, we need to use new instrument transposition information.

**Comment by sammik** [comment]
@oktophonie this should be done, please check, if You have some time

**Review by oktophonie — APPROVED** [review]

**Comment by oktophonie** [comment]
@sammik As far as I can tell this fixes all the issues described, except the cut-and-paste one I mentioned in my comment; as that is not a new issue I am happy to leave it to be fixed separately (otherwise if we're continually adding to this monstrous PR it may never be merged), but if you can incorporate a simple solution here, even better.

The utests will, in any case, need attention before this can be merged.

**Comment by sammik** [comment]
> @sammik  if you can incorporate a simple solution here, even better.
> 
@oktophonie This is not in fact key sig issue. Key sig is just one symptom of problem.

Core problem is, that Instrument change data is lot on paste - it is just simple text pasted (and user needs to right click it and set instrument again to became real Instrument change)

So I propose to adress it as a separate issue.

**Comment by cbjeukendrup** [comment]
@sammik You don't need to worry about providing the translations! As long as the strings are marked as translatable (which is always the case for strings in a .ui file, and in other places you need to wrap them in `qtrc`/`mtrc`/`trc`/`qsTrc` depending on the context), everything will be allright. 

I will review this as soon as I can, but that will take some time!

File: src/engraving/rw/read400/tread.cpp
Comment (RomanPudashkin) [inline src/engraving/rw/read400/tread.cpp:1243]: Reading the new properties must be removed from this reader. We should only read them inside read410

File: src/engraving/libmscore/property.cpp
Comment (RomanPudashkin) [inline src/engraving/libmscore/property.cpp:367]: We use camelCase for property names. I think a better name would be concertKey

File: src/engraving/libmscore/property.cpp
Comment (RomanPudashkin) [inline src/engraving/libmscore/property.cpp:368]: And actualKey instead of t_accidental

File: src/engraving/libmscore/part.h
Comment (RomanPudashkin) [inline src/engraving/libmscore/part.h:46]: I would suggest renaming LESS to AUTO

File: src/engraving/rw/read400/tread.cpp
Comment (RomanPudashkin) [inline src/engraving/rw/read400/tread.cpp:1158]: Looks like a typo

File: src/engraving/libmscore/part.h
Comment (Jojo-Schmitz) [reply src/engraving/libmscore/part.h:46]: Or DEFAULT?

File: src/engraving/rw/read400/tread.cpp
Comment (sammik) [reply src/engraving/rw/read400/tread.cpp:1243]: > Maybe the reason is that we use the 302 reader for unit tests. See `src/engraving/rw/mscloader.cpp`. But I'm not entirely sure if `mscloader` is indeed used for the unit tests.

But it doesnt explain, why if it is in 400 reader, tests pass. Does it?

**Comment by sammik** [comment]
Utests fixed.

Unroll repeats always fails, but I cannot check, whats going on, because unroll repeats is not in MU4 yet - see #9670, so I disabled that test.

File: src/engraving/rw/write/twrite.cpp
Comment (RomanPudashkin) [inline src/engraving/rw/write/twrite.cpp:1792]: These properties should also be renamed to concertKey and actualKey respectively

**Comment by RomanPudashkin** [comment]
@sammik thanks for resolving my comments. We are almost there :)

**Comment by RomanPudashkin** [comment]
@sammik thank you so much!

**Comment by sammik** [comment]
> @sammik thank you so much!

Thank You for cooperation and patience.

Glad, I could help.

**Pull request merged**
